### PR TITLE
FIX: ISXB-704 scroll wheel support windows raw input value

### PR DESCRIPTION
**Pull request by benoitalain**
### Description

Fixes ISXB-704. Based on recently landed [trunk PR](https://github.cds.internal.unity3d.com/unity/unity/pull/50445).

Details of what was changed and the thought process can be found in this Google Document:
https://docs.google.com/document/d/1XCC6GqaWzKf1M1jm-UmKRn6sIlmvF7rvF1C8Ws5Yxa4/edit?usp=sharing

This PR introduces a new Settings option for the InputSystem package. See screenshot:
image

When the "Keep Platform Specific Input Range" option is selected, InputSystem actions that result from a Scroll Wheel hardware input will be expressed in the OS native range when retrieved with the `InputAction.CallbackContext.ReadValue<Vector2>()` method. On Windows, the values should range between -120 to 120, whereas on other platforms it will remain between -1 and 1 in general.

The new setting's default option is "Uniform Across All Platforms". When that option is selected, InputSystem actions that result from a Scroll Wheel will always be expressed in the -1 to 1 range, regardless of platform.

### Changes made

_Please write down a short description of what changes were made._

### Notes

_Please write down any additional notes, remove the section if not applicable._

### Checklist

Before review:

- [ ] Changelog entry added.
    - Explains the change in `Changed`, `Fixed`, `Added` sections.
    - For API change contains an example snippet and/or migration example.
    - FogBugz ticket attached, example `([case %number%](https://issuetracker.unity3d.com/issues/...))`.
    - FogBugz is marked as "Resolved" with *next* release version correctly set.
- [ ] Tests added/changed, if applicable.
    - Functional tests `Area_CanDoX`, `Area_CanDoX_EvenIfYIsTheCase`, `Area_WhenIDoX_AndYHappens_ThisIsTheResult`.
    - Performance tests.
    - Integration tests.
- [ ] Docs for new/changed API's.
    - Xmldoc cross references are set correctly.
    - Added explanation how the API works.
    - Usage code examples added.
    - The manual is updated, if needed.

During merge:

- [ ] Commit message for squash-merge is prefixed with one of the list:
    - `NEW: ___`.
    - `FIX: ___`.
    - `DOCS: ___`.
    - `CHANGE: ___`.
    - `RELEASE: 1.1.0-preview.3`.

After merge:

- [ ] Create forward/backward port if needed. If you are blocked from creating a forward port now please add a task to ISX-1444.

### CODE REVIEW

**Comment by unity-cla-assistant** [comment]
[CLA assistant check](https://cla-assistant.cds.internal.unity3d.com/Unity-Technologies/InputSystem?pullRequest=1946) <br/>All committers have signed the CLA.

File: Packages/com.unity.inputsystem/InputSystem/NativeInputRuntime.cs
Comment (ekcoh) [inline Packages/com.unity.inputsystem/InputSystem/NativeInputRuntime.cs]: Not sure I see what role these getter setters play when earlier Unity version? Wouldn't it make sense to remove the property completely instead?

Comment (ekcoh) [reply Packages/com.unity.inputsystem/InputSystem/NativeInputRuntime.cs]: (or would have expected hard coded behavior, but the answer is - it depends)

File: Packages/com.unity.inputsystem/InputSystem/NativeInputRuntime.cs
Comment (ekcoh) [inline Packages/com.unity.inputsystem/InputSystem/NativeInputRuntime.cs]: Recommend being consistent with getter below and use the short-hand version with =>

File: Packages/com.unity.inputsystem/InputSystem/Plugins/InputForUI/InputSystemProvider.cs
Comment (ekcoh) [inline Packages/com.unity.inputsystem/InputSystem/Plugins/InputForUI/InputSystemProvider.cs:533]: Is InputSystem.scrollWheelDeltaPerTick guaranteed to never be zero (to make sure we never can run into division by zero)

File: Packages/com.unity.inputsystem/InputSystem/Plugins/InputForUI/InputSystemProvider.cs
Comment (ekcoh) [inline Packages/com.unity.inputsystem/InputSystem/Plugins/InputForUI/InputSystemProvider.cs:534]: Nice to avoid the sqr root

File: Packages/com.unity.inputsystem/InputSystem/NativeInputRuntime.cs
Comment (ekcoh) [inline Packages/com.unity.inputsystem/InputSystem/NativeInputRuntime.cs]: I guess we run into issues with UNITY 6 revision prior to trunk change landing but maybe ok if/when that revision is current or is there opportunity for a even more specific define in the asmdef to detect the exact expected version once the trunk change lands maybe?

**Review by ekcoh — COMMENTED** [review]
Commenting for now until aligned on potential graceful version handling.

File: Assets/Tests/InputSystem/Plugins/InputForUITests.cs
Comment (ekcoh) [inline Assets/Tests/InputSystem/Plugins/InputForUITests.cs:498]: Thanks for adding these tests to cover scaling.

**Comment by ekcoh** [comment]
Looks like a branch update and CI rerun might be in order.

File: Packages/com.unity.inputsystem/InputSystem/Plugins/InputForUI/InputSystemProvider.cs
Comment (benoitalain) [reply Packages/com.unity.inputsystem/InputSystem/Plugins/InputForUI/InputSystemProvider.cs:533]: I added an Assert on the native side in the trunk PR, is that enough? I'm not a fan of adding more asserts on all the stages where we propagate that value, do you recommend adding one here too? Only here?

File: Packages/com.unity.inputsystem/InputSystem/NativeInputRuntime.cs
Comment (benoitalain) [reply Packages/com.unity.inputsystem/InputSystem/NativeInputRuntime.cs]: Not sure what the general solution to this problem is. There are other places in the Input System package code where we have a similar problem, so if we find a solution for this we should apply it everywhere (at once) probably.

Comment (ekcoh) [reply Packages/com.unity.inputsystem/InputSystem/NativeInputRuntime.cs]: What I was after above was whether UNITY_6000_0_OR_NEWER was specific enough or whether to define a custom asmdef symbol based on version value to define a symbol more specifically reflecting the change as introduced into UNITY_6000 version. There are other examples here https://github.com/Unity-Technologies/InputSystem/blob/develop/Packages/com.unity.inputsystem/InputSystem/Unity.InputSystem.asmdef but I noticed they very not specific on a "sub-version" level so probably ok.

File: Packages/com.unity.inputsystem/InputSystem/Plugins/InputForUI/InputSystemProvider.cs
Comment (ekcoh) [reply Packages/com.unity.inputsystem/InputSystem/Plugins/InputForUI/InputSystemProvider.cs:533]: If trunk version guarantees that its never zero I believe its fine to not check it in package. Agree that its better to define the API contract on lower level.

**Review by ekcoh — APPROVED** [review]
Approving this now, I saw some minor comments remain but ok with me to leave them as-is if you consider it to be preferable to have it as it currently is.

**Review by duckets — APPROVED** [review]
Approved with amendments to tooltip and API docs

**Review by Pauliusd01 — CHANGES_REQUESTED** [review]
Couple of concerns from me: 

- I did not need these changes to fix the issue, simply opening the Editor version that has your latest changes is enough to fix it
- The ifdefs for the code are not granular enough and I still get errors about missing Editor code on 6.0.6f1 Editor, should that be changed?

Also, please take a look at the slack discussion about this change in our input channel

**Comment by benoitalain** [comment]
> Couple of concerns from me:
> 
> * I did not need these changes to fix the issue, simply opening the Editor version that has your latest changes is enough to fix it
> * The ifdefs for the code are not granular enough and I still get errors about missing Editor code on 6.0.6f1 Editor, should that be changed?
> 
> Also, please take a look at the slack discussion about this change in our input channel

You say you only need the Editor changes, are you sure there's no copy of my package fixes that survived in your Library folder? Otherwise I can't see how the InputSystemUIInputModule fixes could come from the Editor only... makes no sense :-(

About the ifdefs not being granular enough, I'll comment in the slack discussion but I'm not sure there's a clean solution.

**Comment by Pauliusd01** [comment]
> > Couple of concerns from me:
> > 
> > * I did not need these changes to fix the issue, simply opening the Editor version that has your latest changes is enough to fix it
> > * The ifdefs for the code are not granular enough and I still get errors about missing Editor code on 6.0.6f1 Editor, should that be changed?
> > 
> > Also, please take a look at the slack discussion about this change in our input channel
> 
> You say you only need the Editor changes, are you sure there's no copy of my package fixes that survived in your Library folder? Otherwise I can't see how the InputSystemUIInputModule fixes could come from the Editor only... makes no sense :-(

It's the user project from ISXB-704 with no library and 1.8.2 input system

**Comment by benoitalain** [comment]
> > > Couple of concerns from me:
> > > 
> > > * I did not need these changes to fix the issue, simply opening the Editor version that has your latest changes is enough to fix it
> > > * The ifdefs for the code are not granular enough and I still get errors about missing Editor code on 6.0.6f1 Editor, should that be changed?
> > > 
> > > Also, please take a look at the slack discussion about this change in our input channel
> > 
> > 
> > You say you only need the Editor changes, are you sure there's no copy of my package fixes that survived in your Library folder? Otherwise I can't see how the InputSystemUIInputModule fixes could come from the Editor only... makes no sense :-(
> 
> It's the user project from ISXB-704 with no library and 1.8.2 input system

Hi! I've updated the PR description to emphasize the changes that this PR contains. It's not observable scroll range fixes in any of the UI systems, but just the option to allow InputSystem actions returning values in the -120 to 120 range.

**Review by Pauliusd01 — APPROVED** [review]
Thanks for the extra info, lgtm. Do check the CI failures though, there's more than the usual culprit ones

**Pull request merged**
